storage/remote/client.go

```diff
@@ -241,15 +299,35 @@ func (c *Client) Store(ctx context.Context, req []byte, attempt int) error {
 		httpResp.Body.Close()
 	}()
 
-	// TODO-RW11: Here is where we need to handle version downgrade on error
+	// See if we got a X-Prometheus-Remote-Write header in the response
+	if promHeader := httpResp.Header.Get(RemoteWriteVersionHeader); promHeader != "" {
+		// Only update lastRWHeader if the X-Prometheus-Remote-Write header is not blank
+		// (It's blank if it wasn't present, we don't care about that distinction.)
+		c.lastRWHeader = promHeader
+	}
 
 	if httpResp.StatusCode/100 != 2 {
 		scanner := bufio.NewScanner(io.LimitReader(httpResp.Body, maxErrMsgLen))
 		line := ""
 		if scanner.Scan() {
 			line = scanner.Text()
 		}
-		err = fmt.Errorf("server returned HTTP status %s: %s", httpResp.Status, line)
+		switch httpResp.StatusCode {
+		case 400:
+			// Return an unrecoverable error to indicate the 400
+			// This then gets passed up the chain so we can react to it properly
+			// TODO(alexg) Do we want to include the first line of the message?
```

Comment:
Yea, would be nice

Reply:
This is where we could employ more structured response with what series exactly failed.

Reply:
Fixed in https://github.com/prometheus/prometheus/pull/13921